### DISCUSSION TITLE
How to use template, after adding llama_chat_apply_template?

### DISCUSSION THREAD

**Discussion by Floating-Pin** [ggml-org/llama.cpp#5582] [Q&A]
I've been struggling with template for a long time, and now I've discovered that in the last commits https://github.com/ggerganov/llama.cpp/commit/11b12de39bd787c0494da0cd405958fdfedc29c4 what I've been waiting for. But how to use it? Do I need to specify any parameters? Or is everything automatically loaded from the model?

**Comment by ibehnam** [comment]
bumping this because the server docs don't mention this feature.

**Comment by Nick-infinity** [comment]
Any help on this?

**Reply by Floating-Pin** [reply]
Nope.

**Comment by Jeximo** [comment]
> Do I need to specify any parameters? Or is everything automatically loaded from the model?

Hi, Server readme shows command line option [--chat-template JINJA_TEMPLATE](https://github.com/ggerganov/llama.cpp/blob/4e9a7f7f7fb6acbddd1462909c8d696e38edbfcc/examples/server/README.md). By **default**, *llama_chat_apply_template() uses the template from a models metadata*, `tokenizer.chat_template`.

Here's [supported templates for --chat-template](https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template#supported-templates). For example,
```
./server -m ... --chat-template chatml
```

**Reply by ubergarm** [reply]
You can get a list of the ad-hoc implemented chat templates for the /chat/completions llama-server endpoint around this area of the code: https://github.com/ggml-org/llama.cpp/blob/master/src/llama-chat.cpp#L28-L70

The main reason you might want to do this is if the model starts up auto-detecting the wrong chat template. For example, in testing https://huggingface.co/ubergarm/cogito-v2-preview-deepseek-671B-MoE-GGUF for some reason it comes up with what looks like chatml. So running `llama-server .... --chat-template deepseek3` forces it to that internal preset.

Might be something with the jinja template inside the gguf https://huggingface.co/deepcogito/cogito-v2-preview-deepseek-671B-MoE/raw/main/chat_template.jinja containing some extra stuff getting picked up here: https://github.com/ggml-org/llama.cpp/blob/master/src/llama-chat.cpp#L91 or similar.
